src/util/GuardedStream.ts

```diff
@@ -90,13 +72,19 @@ function removeDefaultErrorListener(this: Guarded, event: string): void {
 }
 
 /**
- * Callback that is used to make sure the error-related fallback functions are re-applied
- * when all error listeners are removed.
+ * Makes sure that listeners always receive the error event of a stream,
+ * even if it was thrown before the listener was attached.
+ * If the input is already guarded nothing will happen.
+ * @param stream - Stream that can potentially throw an error.
+ *
+ * @returns The stream.
  */
-attachDefaultErrorListener = function(this: Guarded, event: string): void {
-  if (event === 'error' && this.listenerCount('error') === 0) {
-    this.addListener('error', defaultErrorListener);
-    this.addListener('newListener', removeDefaultErrorListener);
-    this.removeListener('removeListener', attachDefaultErrorListener);
+export function guardStream<T extends NodeJS.EventEmitter>(stream: T): Guarded<T> {
+  const guarded = stream as Guarded<T>;
+  if (!isGuarded(stream)) {
+    guarded[guardedErrors] = [];
+    guarded.on('error', guardingErrorListener);
+    guarded.on('newListener', emitStoredErrors);
```

Comment:
Maybe we want to `prependListener` in both cases, just so it's the first listener?

```suggestion
    guarded.prependListener('error', guardingErrorListener);
    guarded.prependListener('newListener', emitStoredErrors);
```

Reply:
For some reason this breaks the integration test being added here: the logger doesn't get called anymore. Not sure why yet.

Reply:
Never mind then; although I am curious about it.

Reply:
I might have something? Remember the Node.js [code](https://github.com/nodejs/node/blob/053abac02bafafa89824613192a2c7d22950bb01/lib/internal/streams/readable.js#L759-L772) that caused all this.

What happens now when an error is thrown: our error listener sees that there are 2 error listeners attached and does nothing. The Node.js error listener then detaches itself and does not throw an error since there is still an error listener attached. But that would mean that the the Node.js code calls emit again with the same error which should reach our listener, so not sure why that wouldn't trigger then. Unless there is some code somewhere in that whole event chain that makes sure our listener doesn't get the same error twice.

Reply:
Not that I want to spend too much time on this but I was also curious 😄

Reply:
Might be because of the `s.errorEmitted` boolean there. Going to leave at that for now.